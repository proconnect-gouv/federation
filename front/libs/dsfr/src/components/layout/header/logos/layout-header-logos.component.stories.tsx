import { ComponentMeta } from '@storybook/react';

import { withRouter } from '@fc/storybook-config/decorators/with-router';

import {
  LayoutHeaderLogosComponent,
  LayoutHeaderLogosComponentProps,
} from './layout-header-logos.component';

export default {
  component: LayoutHeaderLogosComponent,
  decorators: [withRouter],
  title: 'DSFR/components/layout/header/logos/LayoutHeaderLogosComponent',
} as ComponentMeta<typeof LayoutHeaderLogosComponent>;

export const Default = (args: LayoutHeaderLogosComponentProps) => (
  <LayoutHeaderLogosComponent {...args} />
);
