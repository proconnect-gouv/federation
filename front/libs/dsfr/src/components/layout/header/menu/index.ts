/* istanbul ignore file */

// declarative file
export * from './layout-header-menu.component';
