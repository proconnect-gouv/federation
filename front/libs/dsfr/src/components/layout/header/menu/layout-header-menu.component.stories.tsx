import { ComponentMeta } from '@storybook/react';

import { AppContextProvider } from '@fc/state-management';
import { withRouter } from '@fc/storybook-config/decorators/with-router';

import {
  LayoutHeaderMenuComponent,
  LayoutHeaderMenuComponentProps,
} from './layout-header-menu.component';

export default {
  component: LayoutHeaderMenuComponent,
  decorators: [withRouter],
  title: 'DSFR/components/layout/header/menu/LayoutHeaderMenuComponent',
} as ComponentMeta<typeof LayoutHeaderMenuComponent>;

export const Default = (args: LayoutHeaderMenuComponentProps) => (
  <AppContextProvider
    value={{
      config: {
        OidcClient: {
          endpoints: {
            endSessionUrl: '/endSessionUrl',
            returnButtonUrl: '/returnButtonUrl',
          },
        },
      },
    }}>
    <LayoutHeaderMenuComponent {...args} />
  </AppContextProvider>
);

Default.args = {
  familyName: 'De Gaulle',
  givenName: 'Charles',
  navigationItems: [
    {
      a11y: 'Google',
      href: 'https://www.google.com/',
      label: 'Google',
    },
    {
      a11y: 'DuckDuckGo',
      href: 'https://duckduckgo.com/',
      label: 'DuckDuckGo',
    },
    {
      a11y: 'Qwant',
      href: 'https://www.qwant.com/',
      label: 'Qwant',
    },
  ],
  opened: true,
};
