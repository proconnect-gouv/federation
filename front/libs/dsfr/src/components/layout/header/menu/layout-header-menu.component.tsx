import classnames from 'classnames';
import React, { MouseEventHandler } from 'react';

import { NavigationLink } from '../../../../interfaces';
import { LayoutHeaderToolsComponent } from '../tools';
import { LayoutHeaderNavigationComponent } from './layout-header-navigation.component';

export interface LayoutHeaderMenuComponentProps {
  opened: boolean;
  onClose: MouseEventHandler<HTMLButtonElement>;
  navigationItems?: NavigationLink[];
  lastname: string | undefined;
  firstname: string | undefined;
}

export const LayoutHeaderMenuComponent: React.FC<LayoutHeaderMenuComponentProps> = React.memo(
  ({ firstname, lastname, navigationItems, onClose, opened }: LayoutHeaderMenuComponentProps) => (
    <div
      aria-labelledby="burger-button-mobile-menu"
      className={classnames('fr-header__menu fr-modal', {
        // @NOTE unable to create a eslint rule to match this case
        // eslint-disable-next-line @typescript-eslint/naming-convention
        'fr-modal--opened': opened,
      })}
      id="layout-header-menu-modal">
      <div className="fr-container">
        <button
          aria-controls="layout-header-menu-modal"
          className="fr-btn--close fr-btn"
          onClick={onClose}>
          Fermer
        </button>
        <LayoutHeaderToolsComponent isModalMenu firstname={firstname} lastname={lastname} />
        {navigationItems?.length && (
          <LayoutHeaderNavigationComponent navigationItems={navigationItems} />
        )}
      </div>
    </div>
  ),
);

LayoutHeaderMenuComponent.defaultProps = {
  navigationItems: undefined,
};

LayoutHeaderMenuComponent.displayName = 'LayoutHeaderMenuComponent';
