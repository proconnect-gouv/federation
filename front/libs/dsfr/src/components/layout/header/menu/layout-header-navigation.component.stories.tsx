import { ComponentMeta } from '@storybook/react';

import { withRouter } from '@fc/storybook-config/decorators/with-router';

import {
  LayoutHeaderNavigationComponent,
  LayoutHeaderNavigationComponentProps,
} from './layout-header-navigation.component';

export default {
  component: LayoutHeaderNavigationComponent,
  decorators: [withRouter],
  title: 'DSFR/components/layout/header/menu/LayoutHeaderNavigationComponent',
} as ComponentMeta<typeof LayoutHeaderNavigationComponent>;

export const Default = (args: LayoutHeaderNavigationComponentProps) => (
  <LayoutHeaderNavigationComponent {...args} />
);

Default.args = {
  navigationItems: [
    {
      a11y: 'Google',
      href: 'https://www.google.com/',
      label: 'Google',
    },
    {
      a11y: 'DuckDuckGo',
      href: 'https://duckduckgo.com/',
      label: 'DuckDuckGo',
    },
    {
      a11y: 'Qwant',
      href: 'https://www.qwant.com/',
      label: 'Qwant',
    },
  ],
} as LayoutHeaderNavigationComponentProps;
