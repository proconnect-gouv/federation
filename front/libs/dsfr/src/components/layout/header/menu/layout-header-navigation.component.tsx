import classnames from 'classnames';
import type { MouseEventHandler } from 'react';
import React from 'react';
import { Link, matchPath, useLocation } from 'react-router-dom';

import type { NavigationLink } from '../../../../interfaces';

interface LayoutHeaderNavigationComponentProps {
  navigationItems?: NavigationLink[];
  className?: string;
  onItemClick?: MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>;
}

export const LayoutHeaderNavigationComponent = React.memo(
  ({ className, navigationItems, onItemClick }: LayoutHeaderNavigationComponentProps) => {
    const { pathname } = useLocation();

    return (
      <nav
        aria-label="Menu principal"
        className={classnames('fr-nav', className)}
        id="navigation-869"
        role="navigation">
        <ul className="fr-nav__list">
          {navigationItems &&
            navigationItems.map(({ href: path, label, title }, index) => {
              const isCurrentPage = matchPath(path, pathname);
              const uniqkey = `layout-header-nav-item::${index}`;

              return (
                <li key={uniqkey} className="fr-nav__item">
                  <Link
                    aria-current={isCurrentPage ? 'page' : undefined}
                    className="fr-nav__link"
                    target="_self"
                    title={title}
                    to={path}
                    onClick={onItemClick}>
                    {label}
                  </Link>
                </li>
              );
            })}
        </ul>
      </nav>
    );
  },
);

LayoutHeaderNavigationComponent.displayName = 'LayoutHeaderNavigationComponent';
