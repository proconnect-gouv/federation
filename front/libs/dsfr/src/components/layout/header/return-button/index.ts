/* istanbul ignore file */

// declarative file
export * from './return-button.component';
