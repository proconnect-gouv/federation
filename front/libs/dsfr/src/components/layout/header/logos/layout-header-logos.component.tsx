import React from 'react';
import { Link } from 'react-router-dom';

import { LogoRepubliqueFrancaiseComponent } from '../../../logos';

interface LayoutHeaderLogosComponentProps {
  logo?: string;
  title?: string;
}

export const LayoutHeaderLogosComponent: React.FC<LayoutHeaderLogosComponentProps> = React.memo(
  ({ logo: ApplicationLogo, title }: LayoutHeaderLogosComponentProps): JSX.Element => (
    <React.Fragment>
      <div className="fr-header__logo">
        <Link title={`Accueil - ${title || ''}`} to="/">
          <LogoRepubliqueFrancaiseComponent />
        </Link>
      </div>
      {ApplicationLogo && (
        <div className="fr-header__operator">
          <img
            alt={title}
            className="fr-responsive-img"
            src={ApplicationLogo}
            style={{ height: 54, maxHeight: 54 }}
          />
        </div>
      )}
    </React.Fragment>
  ),
);

LayoutHeaderLogosComponent.defaultProps = {
  logo: undefined,
  title: undefined,
};

LayoutHeaderLogosComponent.displayName = 'LayoutHeaderLogosComponent';
