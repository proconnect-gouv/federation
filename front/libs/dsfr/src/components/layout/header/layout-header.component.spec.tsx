import { render } from '@testing-library/react';
import { useMediaQuery } from 'react-responsive';
import { mocked } from 'ts-jest/utils';

import { AccountContext, AccountInterface } from '@fc/account';
import { AppContextProvider } from '@fc/state-management';

import { LayoutHeaderComponent } from './layout-header.component';
import { LayoutHeaderMobileBurgerButton } from './layout-header-mobile-burger.button';
import { LayoutHeaderLogosComponent } from './logos';
import { LayoutHeaderMenuComponent } from './menu';
import { ReturnButtonComponent } from './return-button';
import { LayoutHeaderToolsComponent } from './tools';

jest.mock('./tools/layout-header-tools.component');
jest.mock('./logos/layout-header-logos.component');
jest.mock('./menu/layout-header-menu.component');
jest.mock('./layout-header-mobile-burger.button');
jest.mock('./return-button/return-button.component');

describe('LayoutHeaderComponent', () => {
  // given
  const navigationItemsMock = [
    { a11y: 'any-a11y-mock-1', href: 'any-href-mock-1', label: 'any-label-mock-1' },
    { a11y: 'any-a11y-mock-2', href: 'any-href-mock-2', label: 'any-label-mock-2' },
  ];

  const accountContextMock = {
    connected: false,
    ready: false,
    userinfos: {
      firstname: expect.any(String),
      lastname: expect.any(String),
    },
  } as unknown as AccountInterface;

  const appContextConfigMock = {
    config: {
      Layout: {
        logo: 'any-logo-mock',
        navigationItems: navigationItemsMock,
      },
      OidcClient: { endpoints: {} },
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    // when
    const { container } = render(
      <AppContextProvider value={appContextConfigMock}>
        <LayoutHeaderComponent />
      </AppContextProvider>,
    );
    // then
    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot, when user is connected', () => {
    // given
    const accountMock = { ...accountContextMock, connected: true, ready: true };
    // when
    const { container } = render(
      <AccountContext.Provider value={accountMock}>
        <AppContextProvider value={appContextConfigMock}>
          <LayoutHeaderComponent />
        </AppContextProvider>
      </AccountContext.Provider>,
    );
    // then
    expect(container).toMatchSnapshot();
  });

  it('should call LayoutHeaderLogosComponent with params', () => {
    // when
    render(
      <AppContextProvider value={appContextConfigMock}>
        <LayoutHeaderComponent />
      </AppContextProvider>,
    );
    // then
    expect(LayoutHeaderLogosComponent).toHaveBeenCalledTimes(1);
    expect(LayoutHeaderLogosComponent).toHaveBeenCalledWith({ logo: 'any-logo-mock' }, {});
  });

  it('should call LayoutHeaderMobileBurgerButton with params when user is connected', () => {
    // given
    const accountMock = { ...accountContextMock, connected: true, ready: true };
    // when
    render(
      <AccountContext.Provider value={accountMock}>
        <AppContextProvider value={appContextConfigMock}>
          <LayoutHeaderComponent />
        </AppContextProvider>
      </AccountContext.Provider>,
    );
    // then
    expect(LayoutHeaderMobileBurgerButton).toHaveBeenCalledTimes(1);
    expect(LayoutHeaderMobileBurgerButton).toHaveBeenCalledWith(
      { onOpen: expect.any(Function), opened: false },
      {},
    );
  });

  it('should call LayoutHeaderToolsComponent with params when user is connected', () => {
    // given
    const accountMock = { ...accountContextMock, connected: true, ready: true };
    // when
    render(
      <AccountContext.Provider value={accountMock}>
        <AppContextProvider value={appContextConfigMock}>
          <LayoutHeaderComponent />
        </AppContextProvider>
      </AccountContext.Provider>,
    );
    // then
    expect(LayoutHeaderToolsComponent).toHaveBeenCalledTimes(1);
    expect(LayoutHeaderToolsComponent).toHaveBeenCalledWith(
      {
        firstname: expect.any(String),
        isDesktopViewport: false,
        isModalMenu: false,
        lastname: expect.any(String),
      },
      {},
    );
  });

  it('should call LayoutHeaderMenuComponent with params when user is connected', () => {
    // given
    const accountMock = { ...accountContextMock, connected: true, ready: true };
    // when
    render(
      <AccountContext.Provider value={accountMock}>
        <AppContextProvider value={appContextConfigMock}>
          <LayoutHeaderComponent />
        </AppContextProvider>
      </AccountContext.Provider>,
    );
    // then
    expect(LayoutHeaderMenuComponent).toHaveBeenCalledTimes(1);
    expect(LayoutHeaderMenuComponent).toHaveBeenCalledWith(
      {
        firstname: expect.any(String),
        lastname: expect.any(String),
        navigationItems: navigationItemsMock,
        onClose: expect.any(Function),
        opened: false,
      },
      {},
    );
  });

  it('should not call ReturnButtonComponent into a mobile viewport, if returnButtonUrl is undefined from OidcClient config', () => {
    // given
    mocked(useMediaQuery).mockReturnValueOnce(true);
    const accountMock = { ...accountContextMock, connected: true, ready: true };
    // when
    render(
      <AccountContext.Provider value={accountMock}>
        <AppContextProvider value={appContextConfigMock}>
          <LayoutHeaderComponent />
        </AppContextProvider>
      </AccountContext.Provider>,
    );
    // then
    expect(ReturnButtonComponent).not.toHaveBeenCalled();
  });

  it('should call ReturnButtonComponent into a mobile viewport, if returnButtonUrl is defined from OidcClient config', () => {
    // given
    const returnButtonUrlMock = 'any-returnButtonUrlMock-mock';
    mocked(useMediaQuery).mockReturnValueOnce(true);
    const accountMock = { ...accountContextMock, connected: true, ready: true };
    // when
    render(
      <AccountContext.Provider value={accountMock}>
        <AppContextProvider
          value={{
            config: {
              Layout: {
                logo: 'any-logo-mock',
                navigationItems: navigationItemsMock,
              },
              OidcClient: { endpoints: { returnButtonUrl: returnButtonUrlMock } },
            },
          }}>
          <LayoutHeaderComponent />
        </AppContextProvider>
      </AccountContext.Provider>,
    );
    // then
    expect(ReturnButtonComponent).toHaveBeenCalledTimes(1);
    expect(ReturnButtonComponent).toHaveBeenCalledWith(
      {
        isMobileViewport: true,
        url: returnButtonUrlMock,
      },
      {},
    );
  });
});
