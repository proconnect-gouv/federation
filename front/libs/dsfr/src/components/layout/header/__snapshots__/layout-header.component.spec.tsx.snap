// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutHeaderComponent should match the snapshot 1`] = `
<div>
  <header
    class="fr-header"
    role="banner"
  >
    <div
      class="fr-header__body"
    >
      <div
        class="fr-container"
      >
        <div
          class="fr-header__body-row"
        >
          <div
            class="fr-header__brand fr-enlarge-link"
          >
            <div
              class="fr-header__brand-top"
            >
              <div>
                LayoutHeaderLogosComponent
              </div>
            </div>
          </div>
          <div
            class="fr-header__tools"
          >
            <div>
              LayoutHeaderToolsComponent
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`LayoutHeaderComponent should match the snapshot, when user is connected 1`] = `
<div>
  <header
    class="fr-header"
    role="banner"
  >
    <div
      class="fr-header__body"
    >
      <div
        class="fr-container"
      >
        <div
          class="fr-header__body-row"
        >
          <div
            class="fr-header__brand fr-enlarge-link"
          >
            <div
              class="fr-header__brand-top"
            >
              <div>
                LayoutHeaderLogosComponent
              </div>
              <div>
                LayoutHeaderMobileBurgerButton
              </div>
            </div>
          </div>
          <div
            class="fr-header__tools"
          >
            <div>
              LayoutHeaderToolsComponent
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      LayoutHeaderMenuComponent
    </div>
  </header>
</div>
`;
